docs: clarify vim.system and fix style

telemachus opened the pull request.

telemachus commented:
This seem too minor to bother with, but it would have helped me.

Quick background, I came to `uv.spawn` from `vim.system`, and I
came to `vim.system` from `jobstart`. `jobstart` does not error
if a command fails to start; instead it returns 0 or -1. I wasn't
sure what `vim.system` would do in case of failure to run a
command, and `uv.spawn` didn't make that clear (to me). This one
sentence would have helped.

gpanders commented:
Thanks! I’m adding the typo label so that this change will be hatched with other small changes. cc @dundargoc

zeertzjq commented:
Shouldn't this change be applied to <https://github.com/luvit/luv/blob/master/docs.md>?

dundargoc commented:
Yeah, since we copy the luv docs verbatim.

@telemachus mind sending patch to luv instead? If they merge it then it'll end up in neovim when we bump luv version.

telemachus commented:
Done at https://github.com/luvit/luv/pull/679

For what it's worth, I can see several advantages to tying neovim's docs to 
luv's docs, but the result makes me wish that neovim's docs for vim.system 
were less brief.

As it stands, some people (like me) will probably end up at vim.system because 
jobstart says "Prefer vim.system in Lua." The docs at vim.system show a single
(very simple) example and then say, "See uv.spawn for more details."

As I understand it, vim.system is meant for people (again, like me) who only 
need a very limited range of what luv offers. Rather than make those people read 
the entire luv documentation, maybe vim.system should say more (and/or offer more
examples)?

tl;dr — how would you all feel about a PR to expand the documentation for 
vim.system instead of uv.spawn? (The vim.system docs are not tied to anything 
else, right?)

clason commented:
If there's specific changes that you think can improve the `vim.system` docs, go ahead. But as a general principle we do not want to duplicate information across the documentation, as this makes it much harder to keep things consistent. So making the `vim.system` docs "standalone" is not desired.

telemachus commented:
Fair enough. I'll wait and see if luv accepts my PR. If they don't, I'll submit something
here, specifying that `vim.system` will throw an error on failure to run a command.

clason commented:
They will, and we'll port the change. Might just take a while.

telemachus commented:
So, it looks like we are back to square one. The luv library is not throwing the error: neovim is. (See conversation [here and following](https://github.com/luvit/luv/pull/679#issuecomment-1783792431).)

If neovim wraps luv and alters its behavior, then it doesn't seem right to simply copy luv's documentation. I don't know what the best way forward is, but I don't think users of `vim.system` or `uv.spawn` should need to read [neovim's source code](https://github.com/neovim/neovim/blob/7a5effb0f95e295c265fe09e7414d859a6d79657/runtime/lua/vim/_system.lua#L241) or use trial and error in order to find out that these functions can throw an error.

I'm happy to work on a PR if a fix would be accepted.

clason commented:
No, if it's Neovim behavior, it obviously needs to be documented in the corresponding Neovim doc.

gpanders commented:
> So, it looks like we are back to square one. The luv library is not throwing the error: neovim is. (See conversation [here and following](https://github.com/luvit/luv/pull/679#issuecomment-1783792431).)

Oops! That's my mistake. Apologies for steering you in the wrong direction.

Re-opening this in light of this information.

clason reviewed (changes requested):
That information needs to go into the `vim.system` docstring then, not the `luv` reference!

telemachus commented:
> > So, it looks like we are back to square one. The luv library is not throwing the error: neovim is. (See conversation [here and following](https://github.com/luvit/luv/pull/679#issuecomment-1783792431).)
> 
> Oops! That's my mistake. Apologies for steering you in the wrong direction.

@gpanders No worries. I appreciate everyone's help so far.

> That information needs to go into the vim.system docstring then, not the luv reference!

@clason Okay, cool. That seems to mean I want to edit `runtime/lua/vim/_editor.lua`. There is a `TODO` in there already for `vim.system`:

```lua
-- TODO(lewis6991): document that the signature is system({cmd}, [{opts},] {on_exit})
```

I think that TODO has been there since `vim.system` [was first added](https://github.com/neovim/neovim/commit/c0952e62fd0ee16a3275bb69e0de04c836b39015#diff-79965f18e45c27a2d629630a82c115f9fb93ccfb1c7ed9e6a42125e56a8bf48d). Can I try to fix that as well or would you recommend I leave it alone?

telemachus committed:
docs: clarify vim.system and style fixes

+ Explicitly mention that vim.system throws an error (unlike uv.spawn)
+ Use present tense rather than imperatives consistently

telemachus commented:
> That information needs to go into the `vim.system` docstring then, not the `luv` reference!

@clason Done. Two notes:

1. In this version, I stated that `vim.system` can throw an error twice. You may want only one of these to remain, but this way you can decide which placement is better.
2. [`:help dev-doc`] requests present tense rather than the imperative. This file was a mix of the two, so I tried to make everything present tense.

gpanders commented on runtime/lua/vim/_editor.lua:
```diff
@@ -82,16 +82,17 @@ vim.log = {
 ---
---- See |uv.spawn()| for more details.
+--- See |uv.spawn()| for more details. Note: unlike |uv.spawn()|, vim.system
+--- throws an error if the cmd cannot be run.
```
```suggestion
--- throws an error if {cmd} cannot be run.
```

telemachus replied:
Done

gpanders commented:
I re-added the `typo` label as these changes are small enough to be batched with the other typo changes.

dundargoc commented:
Thank you for your contribution! We collect all typo fixes and similar small fixes into a single pull request and merge it once it gets big enough: https://github.com/neovim/neovim/pull/25831
